kucherov_vasilii/graph_json_printer.cpp

```diff
@@ -0,0 +1,182 @@
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+const int kVerticesCount = 14;
+
+class Graph {
+ public:
+  using VertexId = int;
+  using EdgeId = int;
+
+  struct Edge {
+   public:
+    Edge(EdgeId id, VertexId from_vertex_id, VertexId to_vertex_id)
+        : id_(id),
+          from_vertex_id_(from_vertex_id),
+          to_vertex_id_(to_vertex_id) {}
+
+    EdgeId id() const { return id_; }
+    VertexId from_vertex_id() const { return from_vertex_id_; }
+    VertexId to_vertex_id() const { return to_vertex_id_; }
+
+   private:
+    EdgeId id_ = 0;
+    VertexId from_vertex_id_ = 0;
+    VertexId to_vertex_id_ = 0;
+  };
+
+  struct Vertex {
+   public:
+    explicit Vertex(VertexId id) : id_(id) {}
+    VertexId id() const { return id_; }
+
+   private:
+    VertexId id_ = 0;
+  };
+
+  void add_vertex() { vertices_.emplace_back(get_new_vertex_id()); }
+
+  void add_edge(VertexId from_vertex_id, VertexId to_vertex_id) {
+    edges_.emplace_back(get_new_edge_id(), from_vertex_id, to_vertex_id);
+  }
+
+  std::vector<EdgeId> get_vertex_edges_ids(const Graph::Vertex& vertex) const {
+    std::vector<EdgeId> edge_ids;
+
+    for (const auto& edge : edges_) {
+      if (edge.from_vertex_id() == vertex.id() ||
+          edge.to_vertex_id() == vertex.id()) {
+        edge_ids.push_back(edge.id());
+      }
+    }
+
+    return edge_ids;
+  }
```

Comment:
1) `vertex` - избыточная информация, как вы видите, в методе везде используется только `vertex.id()`, следовательно какие данные нужны этому методу?
2) Название кривоватое, давайте лучше например `get_connected_edge_ids`
3) Обходить все вершины в массиве довольно медленно, особенно если их будет 100500 штук, лучше добавьте лист связности, например `std::unordered_map<VertexId, std::vector<EdgeId>>adjacency_list_`

Reply:
```suggestion
  const std::vector<EdgeId>& get_connected_edge_ids(VertexId vertex_id) const {
    return adjacency_list_.at(vertex_id);
  }
```
Переписал `get_vertex_edges_ids` в такую штуку. Вопрос по методу `unordered_map.at()`, судя по документации, он выбросит исключение, если не найдет переданный ключ. С другой стороны возвращать пустой вектор в таком случае неверно, так как в графе могут быть просто вершины без ребер. Собственно вопрос - допустима ли такая реализация, или выброс исключения тоже не очень хорошо?

Reply:
Ниже ответил